### ISSUE TITLE
Java-lang integration tests fail

### ISSUE THREAD

**Issue by Shorttail** [gradle/gradle#2436]
### Expected Behavior
Integration tests run through `java-lang`, `jvm-component`, and `junit-test-suite` same way unit tests do.

### Current Behavior
Integration tests fail with a `ClassNotFoundException`.

### Context
Using Gradle 3.5, the integration tests ran fine. Using Gradle 4.0.1 the tests all fail the same way.

### Steps to Reproduce (for bugs)
Using the above mentioned plugins, under `testSuites`, create a test `JUnitTestSuiteSpec` that is NOT named `test`, such as `integrationTest`. Place any JUnit test in the folder (`src/integrationTest/java/...`). When run through `build` it will fail with a `ClassNotFoundException`.

Considering all tests in `test` folders run and all tests in `integrationTest` folders fail, I'm guessing the plugin simply forgets that it should search for the class files in the `integrationTest` folders. The same tests configured with the mature `java` plugin work just fine when configured as follows:
```gradle
sourceSets {
	test {
		java.srcDirs = ['src/test/java']
		resources.srcDirs = ['src/test/resources']
	}
	
	integrationTest {
		java.srcDirs = ['src/integrationTest/java']
		resources.srcDirs = ['src/integrationTest/resources']
		compileClasspath += main.output + test.output
		runtimeClasspath += main.output + test.output
	}
}

task integrationTest(type: Test) {
	setTestClassesDirs sourceSets.integrationTest.output.classesDirs
	classpath = sourceSets.integrationTest.runtimeClasspath
}
```

### Your Environment

Gradle 4.0.1, Windows 10 10.0, VM is Zulu 1.8.0_131-b11, x86_64 Ryzen 1700 CPU, 16 GB RAM.

Stacktrace:

```
java.lang.ClassNotFoundException: example.com.Test
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:66)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy1.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by bmuschko** [comment]
What exactly is `testSuites`? Could you please provide a simplistic example in a GitHub repository that reproduces the issue?

**Comment by Shorttail** [comment]
`testSuites` is part of the `junit-test-suite` DSL:

	model {
		components {
			main(JvmLibrarySpec)
		}
		
		testSuites {
			test(JUnitTestSuiteSpec) {
				jUnitVersion 4.12
				testing $.components.main
			}
		}
	}

I tried to make a minimum example, but I can't reproduce the error. However, updating to Gradle 4.1 my main project no longer fails finding the integration test classes, so I guess the issue was fixed somewhere.

**Comment by bmuschko** [comment]
How much does your project lean on the software model? We haven't made much progress on the that end. I'd actually suggest you go without it for building a Java-based project. It would be helpful if you could provide a simplistic example on GitHub.

**Comment by Shorttail** [comment]
>How much does your project lean on the software model?

My project uses the old Java plugin and the new software model in parallel, because other plugins don't work at all with the new software model, such as the Eclipse plugin (projects don't have the Java type). I'll consider scrapping it though.

>It would be helpful if you could provide a simplistic example on GitHub.

I managed to reproduce the bug just now:
https://github.com/Shorttail/gradle-issue-2436

**Comment by bmuschko** [comment]
> I managed to reproduce the bug just now.

Thanks for providing the sample. I can reproduce it with 4.0.1 and 4.0.2. However, it is fixed with 4.1. I suggest you go with that version. I couldn't really identify the fix in the [4.1 release notes](https://docs.gradle.org/current/release-notes.html#fixed-issues).
